Gateway Configuration Server configuration endpoint

adriansmares opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

**Summary:**
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes #64
References #68 

**Changes:**
<!-- What are the changes made in this pull request? -->

- Added the `/api/v3/gcs/gateways/:gateway_id/global_conf.json` that generates the gateway configuration.

**Notes for Reviewers:**
<!--
Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

Unregistered gateways cannot be retrieved.

**Release Notes: _(optional)_**
<!--
Any notes that we need to include in the Release Notes for the next release.
These notes are formatted as bullet points, written in past tense, and will be
combined with the labels of this Pull Request.
- Always mention changes in API, database models, configuration options or defaults.
- Are there any database migrations required?
- What are the functional or behavioral changes?
-->

- Added the `/api/v3/gcs/gateways/<gateway ID>/global_conf.json` that generates the gateway configuration in the JSON format based on the frequency plan.

johanstokking commented on cmd/internal/shared/errors.go:
```diff
@@ -25,4 +25,5 @@ var (
 	ErrInitializeApplicationServer = errors.Define("initialize_application_server", "could not initialize Application Server")
 	ErrInitializeJoinServer        = errors.Define("initialize_join_server", "could not initialize Join Server")
 	ErrInitializeConsole           = errors.Define("initialize_console", "could not initialize Console")
+	ErrInitializeGCS               = errors.Define("initialize_gateway_configuration_server", "could not initialize Gateway Configuration Server")
```
GCS is written out above. It's a bit long, but still

johanstokking commented on cmd/ttn-lw-stack/commands/config.go:
```diff
@@ -43,7 +44,8 @@ type Config struct {
 	GCS              struct {
-		BasicStation cups.ServerConfig `name:"basic-station"`
+		gatewayconfigurationserver.Config `name:",squash"`
+		BasicStation                      cups.ServerConfig `name:"basic-station"`
```
Can we make Basic Station part of the `gatewayconfigurationserver` config? Or what is the reasoning for this cohesion?

adriansmares replied:
The initial thinking was that each component is started separately by the stack binary instead of the GCS itself. I can see why this is inconsistent, and have corrected it accordingly.

johanstokking commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver_test.go:
```diff
@@ -0,0 +1,140 @@
+		ctx := c.FillContext(c.Context())
+		gcs, err := New(c, testConfig, WithRegistry(gs), WithContext(newContextWithRightsFetcher(ctx)))
+		_ = gcs
+		test.Must(nil, err)
```
Check `should.IsNotNil` and fatal otherwise to avoid panics

johanstokking commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver_test.go:
```diff
@@ -0,0 +1,140 @@
+)
+
+func TestWeb(t *testing.T) {
+	t.Run("GetGateway", func(t *testing.T) {
```
Are you planning other tests in this func? Because otherwise you don't need this

adriansmares replied:
Not planned for now. Removed.

johanstokking commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver_test.go:
```diff
@@ -0,0 +1,140 @@
+		_ = gcs
+		test.Must(nil, err)
+		test.Must(nil, c.Start())
+		defer c.Close()
```
Maybe add some blank lines above to keep things a bit separate

htdvisser commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver.go:
```diff
@@ -0,0 +1,155 @@
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/pfconfig/semtechudp"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	ttnweb "go.thethings.network/lorawan-stack/pkg/web"
```
What other web is there? Don't alias things if not needed.

htdvisser commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver.go:
```diff
@@ -0,0 +1,155 @@
+	gcs := &GatewayConfigurationServer{
+		Component: c,
+		config:    conf,
+		ctx:       c.FillContext(c.Context()),
```
Is this necessary? I thought the component's context already had everything filled that needed to be filled. As far as I know the `FillContext` was meant for filling request contexts that didn't pass the interceptor/echo middleware. cc: @johanstokking

htdvisser commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+// WithContext sets the context for the given server.
+func WithContext(ctx context.Context) Option {
```
Why would we replace the context? We should inherit it from the Component, right?

adriansmares replied:
I'm using it inside the tests - It is required for the `rights` fetcher.

johanstokking replied:
We shouldn't be doing that. If you need a rights fetcher, consider spinning up a mock IS, or make sure a fetcher is available in the context where you need it.

johanstokking replied:
This is not addressed

htdvisser commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver.go:
```diff
@@ -0,0 +1,155 @@
+	}
+}
+
+func (gcs *GatewayConfigurationServer) handleGet(c echo.Context) error {
```
handleGet is too vague, maybe rename to handleGetGlobalConfigJSON or something

htdvisser commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver.go:
```diff
@@ -0,0 +1,155 @@
+)
+
+func (gcs *GatewayConfigurationServer) getGateway(ctx context.Context, ids ttnpb.GatewayIdentifiers, fieldmask types.FieldMask) (*ttnpb.Gateway, error) {
+	ctx = gcs.FillContext(ctx)
```
`getGateway` is only called by `handleGet` which already called `getContext` that calls `FillContext`.

htdvisser commented on pkg/gatewayconfigurationserver/middleware.go:
```diff
@@ -0,0 +1,86 @@
+	"google.golang.org/grpc/metadata"
+)
+
+func (gcs *GatewayConfigurationServer) handleError() echo.MiddlewareFunc {
```
Why do we keep adding new error handling middlewares? The idea was to have only one so that we return consistent responses and don't have to make changes in 10 places.

htdvisser replied:
Still open

htdvisser commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver.go:
```diff
@@ -0,0 +1,155 @@
+
+func (gcs *GatewayConfigurationServer) getGateway(ctx context.Context, ids ttnpb.GatewayIdentifiers, fieldmask types.FieldMask) (*ttnpb.Gateway, error) {
+	ctx = gcs.FillContext(ctx)
+	if ids.IsZero() {
```
Isn't this already caught by the middleware?

johanstokking commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver.go:
```diff
@@ -0,0 +1,144 @@
+	gcs := &GatewayConfigurationServer{
+		Component: c,
+		config:    conf,
+		ctx:       c.Context(),
```
We don't need this; the context is available through `Component.Context()` at any time

johanstokking commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver.go:
```diff
@@ -0,0 +1,144 @@
+
+func (gcs *GatewayConfigurationServer) getRegistry(ctx context.Context) ttnpb.GatewayRegistryClient {
+	if gcs.registry != nil {
+		return gcs.registry
```
Is this also for testing?

johanstokking commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver_test.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+func init() {
+	testConfig.TheThingsGateway.Default.FirmwareURL = "http://example.com"
```
Why here?

johanstokking commented:
@adriansmares note that you can add a context filler from the test, and that you can simply fill the context when handling the request. That way, you can get arbitrary testing stuff from tests in your handler, without options

adriansmares commented:
I've made the requested changes regarding testing: we no longer use the forced contexts / gateway registry clients. I've converted the old mock gateway registry to a mock IS instead, and used the context-filler suggestion for the rights hook.

johanstokking commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver.go:
```diff
@@ -0,0 +1,113 @@
+
+	"github.com/gogo/protobuf/types"
+	echo "github.com/labstack/echo/v4"
+	bs_cups "go.thethings.network/lorawan-stack/pkg/basicstation/cups"
```
I don't think underscores in imports are what we typically do

htdvisser replied:
There isn't really a guideline for that. The gRPC plugins/middlwares we use also have underscores (even in their package name). Using `bscups` is fine here.

johanstokking commented on pkg/gatewayconfigurationserver/gatewayconfigurationserver.go:
```diff
@@ -0,0 +1,113 @@
+		middleware = append(middleware, gcs.requireGatewayRights(ttnpb.RIGHT_GATEWAY_INFO))
+	}
+	group := server.Group(ttnpb.HTTPAPIPrefix+"/gcs/gateways/:gateway_id", middleware...)
+	group.GET("/global_conf.json", gcs.handleGetGlobalConfig)
```
Can we put this in a group indicating that this is for Semtech UDP packet forwarders? I.e. `gcs/gateways/:gateway_id/semtechudp/global_conf.json`

htdvisser commented on cmd/internal/shared/gatewayconfigurationserver/config.go:
```diff
@@ -0,0 +1,27 @@
+)
+
+// DefaultGatewayConfigurationServerConfig is the default configuration for the Application Server.
+var DefaultGatewayConfigurationServerConfig = gatewayconfigurationserver.Config{}
```
I think we should `RequireAuth: true` here. Both the basicstation and The Things Gateway can do auth just fine.

htdvisser commented on pkg/gatewayconfigurationserver/middleware.go:
```diff
@@ -0,0 +1,86 @@
+	}
+}
+
+const (
```
Why the ()?

htdvisser commented on pkg/gatewayconfigurationserver/middleware.go:
```diff
@@ -0,0 +1,86 @@
+			gtwID := c.Get(gatewayIDKey).(ttnpb.GatewayIdentifiers)
+			md := metadata.New(map[string]string{
+				"id":            gtwID.GatewayID,
+				"authorization": c.Request().Header.Get(echo.HeaderAuthorization),
```
This (and the lines below) is already done in `gcs.getContext`

htdvisser commented on pkg/thethingsgateway/cups/middleware.go:
```diff
@@ -0,0 +1,58 @@
+var errUnauthenticated = errors.DefineUnauthenticated("unauthenticated", "call was not authenticated")
+
+// validateAndFillGatewayIDs checks if the request contains a valid gateway ID.
+func (s *Server) validateAndFillGatewayIDs() echo.MiddlewareFunc {
```
Can't we just assume that `gcs.validateAndFillIDs` was already in the chain?

htdvisser replied:
https://github.com/TheThingsNetwork/lorawan-stack/issues/583

htdvisser commented on pkg/thethingsgateway/cups/middleware.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+// checkAuthPresence checks if the request contains the Authorization header.
+func (s *Server) checkAuthPresence() echo.MiddlewareFunc {
```
Let's rename this to `requireAuth`. It doesn't just check, it also errors if it's not there.

adriansmares commented:
I've added the technical debt issue to #583 and fixed the comments. Regarding the middleware (since for some reason I cannot comment directly there) - I've forgot to remove the function, but it was already removed from the chain, sorry.
